datafusion-examples/examples/async_udf.rs

```diff
@@ -0,0 +1,405 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow_schema::{Field, Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::{
+    arrow::{
+        array::{ArrayRef, Float32Array, Float64Array},
+        datatypes::DataType,
+        record_batch::RecordBatch,
+    },
+    execution::{
+        context::{QueryPlanner, SessionState},
+        runtime_env::RuntimeEnv,
+        TaskContext,
+    },
+    logical_expr::Volatility,
+    physical_expr::PhysicalSortExpr,
+    physical_plan::{
+        stream::RecordBatchStreamAdapter, DisplayFormatType, Distribution, ExecutionPlan,
+        Partitioning, SendableRecordBatchStream, Statistics,
+    },
+    physical_planner::{DefaultPhysicalPlanner, ExtensionPlanner, PhysicalPlanner},
+};
+
+use datafusion::prelude::*;
+use datafusion::{error::Result, physical_plan::functions::make_scalar_function};
+use datafusion_common::{
+    cast::{as_float32_array, as_float64_array},
+    tree_node::{Transformed, TreeNode},
+    DFSchemaRef, DataFusionError,
+};
+use datafusion_expr::{
+    expr::ScalarUDF, Extension, LogicalPlan, Subquery, UserDefinedLogicalNode,
+    UserDefinedLogicalNodeCore,
+};
+use datafusion_optimizer::{optimize_children, OptimizerConfig, OptimizerRule};
+use std::{
+    any::Any,
+    fmt::{self, Debug},
+    sync::Arc,
+};
+
+use futures::{FutureExt, StreamExt};
+
+// create local execution context with an in-memory table
+fn create_context() -> Result<SessionContext> {
+    // define a schema.
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("a", DataType::Float32, false),
+        Field::new("b", DataType::Float64, false),
+    ]));
+
+    // define data.
+    let batch = RecordBatch::try_new(
+        schema,
+        vec![
+            Arc::new(Float32Array::from(vec![2.1, 3.1, 4.1, 5.1])),
+            Arc::new(Float64Array::from(vec![1.0, 2.0, 3.0, 4.0])),
+        ],
+    )?;
+
+    // declare a state with a query planner and an optimizer rule
+    let config = SessionConfig::new();
+    let runtime = Arc::new(RuntimeEnv::default());
+    let state = SessionState::with_config_rt(config, runtime)
+        .with_query_planner(Arc::new(PowQueryPlanner {}))
+        .add_optimizer_rule(Arc::new(PowOptimizerRule {}));
+
+    // declare a new context. In spark API, this corresponds to a new spark SQLsession
+    let ctx = SessionContext::with_state(state);
+
+    // declare a table in memory. In spark API, this corresponds to createDataFrame(...).
+    ctx.register_batch("t", batch)?;
+    Ok(ctx)
+}
+
+// pow is similar to the pow function in simple_udf example
```

Comment:
I think the example is already complex and long. I would also like to avoid dependencies on IO, especially network.
What do you think about sending data from main to the UDF via a channel? Tokio sync is already a dev dependency.